File: auth-api/src/main/java/com/cartobucket/auth/api/server/routes/GroupMembers.java
/*
 * Copyright 2023 Bryce Groff (Cartobucket LLC)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.cartobucket.auth.api.server.routes;

import com.cartobucket.auth.data.domain.Page;
import com.cartobucket.auth.api.interfaces.GroupMembersApi;
import com.cartobucket.auth.api.dto.GroupMemberRequest;
import com.cartobucket.auth.api.dto.GroupMembersResponse;
import com.cartobucket.auth.api.server.routes.mappers.GroupMemberMapper;
import com.cartobucket.auth.data.services.GroupMemberService;
import jakarta.ws.rs.core.Response;

import java.util.Collections;
import java.util.UUID;

import static com.cartobucket.auth.api.server.routes.Constants.LIMIT_DEFAULT;
import static com.cartobucket.auth.api.server.routes.Constants.OFFSET_DEFAULT;

public class GroupMembers implements GroupMembersApi {
    final GroupMemberService groupMemberService;

    public GroupMembers(GroupMemberService groupMemberService) {
        this.groupMemberService = groupMemberService;
    }

    @Override
    public Response createGroupMember(UUID groupId, GroupMemberRequest groupMemberRequest) {
        return Response
                .ok()
                .entity(
                        GroupMemberMapper.toResponse(
                                groupMemberService.createGroupMember(GroupMemberMapper.from(groupId, groupMemberRequest))
                        )
                )
                .build();
    }

    @Override
    public Response deleteGroupMember(UUID groupId, UUID memberId) {
        groupMemberService.deleteGroupMember(memberId);
        return Response
                .noContent()
                .build();
    }

    @Override
    public Response getGroupMember(UUID groupId, UUID memberId) {
        return Response
                .ok()
                .entity(
                        GroupMemberMapper.toResponse(groupMemberService.getGroupMember(memberId))
                )
                .build();
    }

    @Override
    public Response listGroupMembers(UUID groupId, Integer limit, Integer offset) {
        if (limit == null) {
            limit = LIMIT_DEFAULT;
        }
        if (offset == null) {
            offset = OFFSET_DEFAULT;
        }

        final var groupMembersResponse = new GroupMembersResponse();
        groupMembersResponse.setGroupMembers(
                groupMemberService
                        .getGroupMembers(Collections.emptyList(), Collections.singletonList(groupId), new Page(limit, offset))
                        .stream()
                        .map(GroupMemberMapper::toResponse)
                        .toList()
        );
        groupMembersResponse.setPage(getGroupMembersPage(groupId, limit, offset));
        return Response
                .ok()
                .entity(groupMembersResponse)
                .build();
    }

    private com.cartobucket.auth.api.dto.Page getGroupMembersPage(UUID groupId, Integer limit, Integer offset) {
        var page = new com.cartobucket.auth.api.dto.Page();
        var next = "/groups/" + groupId + "/members?limit=" + limit + "&offset=" + (offset + limit);
        page.setNext(next);
        if (offset > 0) {
            var previous = "/groups/" + groupId + "/members?limit=" + limit + "&offset=" + (offset - limit);
            page.setPrevious(previous);
        }
        return page;
    }
}

File: web/src/main/kotlin/com/revethq/auth/web/api/routes/GroupMembers.kt
/*
 * Copyright 2023 Bryce Groff (Revet)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.revethq.auth.web.api.routes

import com.revethq.auth.core.api.dto.GroupMemberRequest
import com.revethq.auth.core.api.dto.GroupMembersResponse
import com.revethq.auth.core.api.dto.Page as PageDto
import com.revethq.auth.core.api.interfaces.GroupMembersApi
import com.revethq.auth.web.api.routes.Constants.LIMIT_DEFAULT
import com.revethq.auth.web.api.routes.Constants.OFFSET_DEFAULT
import com.revethq.auth.web.api.routes.mappers.GroupMemberMapper
import com.revethq.auth.core.domain.Page
import com.revethq.auth.core.services.GroupMemberService
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.ws.rs.core.Response
import java.util.UUID

@ApplicationScoped
open class GroupMembers @Inject constructor(
    private val groupMemberService: GroupMemberService
) : GroupMembersApi {

    override fun createGroupMember(groupId: UUID, groupMemberRequest: GroupMemberRequest): Response {
        return Response
            .ok()
            .entity(
                GroupMemberMapper.toResponse(
                    groupMemberService.createGroupMember(GroupMemberMapper.from(groupId, groupMemberRequest))
                )
            )
            .build()
    }

    override fun deleteGroupMember(groupId: UUID, memberId: UUID): Response {
        groupMemberService.deleteGroupMember(memberId)
        return Response.noContent().build()
    }

    override fun getGroupMember(groupId: UUID, memberId: UUID): Response {
        return Response
            .ok()
            .entity(GroupMemberMapper.toResponse(groupMemberService.getGroupMember(memberId)))
            .build()
    }

    override fun listGroupMembers(groupId: UUID, limit: Int?, offset: Int?): Response {
        val actualLimit = limit ?: LIMIT_DEFAULT
        val actualOffset = offset ?: OFFSET_DEFAULT

        val groupMembersResponse = GroupMembersResponse()
        groupMembersResponse.groupMembers = groupMemberService
            .getGroupMembers(emptyList(), listOf(groupId), Page(actualLimit, actualOffset))
            .map { GroupMemberMapper.toResponse(it) }
        groupMembersResponse.page = getGroupMembersPage(groupId, actualLimit, actualOffset)
        return Response.ok().entity(groupMembersResponse).build()
    }

    private fun getGroupMembersPage(groupId: UUID, limit: Int, offset: Int): PageDto {
        val page = PageDto()
        val next = "/groups/$groupId/members?limit=$limit&offset=${offset + limit}"
        page.next = next
        if (offset > 0) {
            val previous = "/groups/$groupId/members?limit=$limit&offset=${offset - limit}"
            page.previous = previous
        }
        return page
    }
}
